src/runner.rs

```diff
@@ -1263,8 +1260,47 @@ impl ValueStack {
 		Ok(())
 	}
 
+	#[inline]
+	fn extend(&mut self, values: &[RuntimeValue]) -> Result<(), TrapKind> {
+		let len = cmp::min(values.len(), self.buf.len() - self.sp);
+		self.buf[self.sp..(self.sp+len)].copy_from_slice(&values[..len]);
+		self.sp += len;
+		if len < values.len() {
+			Err(TrapKind::StackOverflow)
```

Comment:
I wonder why do we return `StackOverflow` here, after we've tried to perform copy (even though it is copy of zero elements) and we've modified the stack pointer?

It's actually violates consistency of the stack. Consider this test:


```rust
let values = vec![
    RuntimeValue::I32(1), 
    RuntimeValue::I32(2), 
    RuntimeValue::I32(3),
    RuntimeValue::I32(4), 
    RuntimeValue::I32(5),
];

let mut stack = ValueStack::with_limit(4);
assert!(stack.extend(&values).is_err());
assert!(stack.push(RuntimeValue::I32(0)).is_ok());
```

this test is going to fail since `sp` will be modified and will point out of the underlying buffer.

Reply:
I think we need a bit clarification on the semantics of the api, I was trying to follow the previous code's semantics:

```
               for local in locals {
                       value_stack.push(local)
                               .map_err(|_| TrapKind::StackOverflow)?;
               }
```

This code will insert `locals' until it overflows, when it overflows, all previous entries have been pushed into value_stack already. That's why I do a copy first and then if we figure out we're not able to copy all, then simply return a 'overflow' error.